nucypher/acumen/perception.py

```diff
@@ -218,41 +218,91 @@ def abridged_state_details(state):
+        if layer is None:
+            existing_strategies.append(strategy)
+        else:
+            existing_strategies.insert(strategy, index=layer)
+
     def get_nodes(self, label=None) -> Iterator["Teacher"]:
-        """If label is None return all"""
+        """If label is None return all know nodes"""
         if not label:
             return iter(self)
         if label not in self.BUCKETS:
             raise self.UnknownLabel(f'{label} is not a valid node label')
         try:
-            nodes = self._marked[label]
+            nodes = self.__marked[label]
         except KeyError:
             return iter(dict())  # empty
         return iter(nodes)
 
-    def mark_as(self, label, node: "Teacher") -> None:
-
+    def label(self, label, node: "Teacher") -> None:
+        """
+        Apply a label to a known node.  Removes any exiting labels before adding the new one.
+        If the provided label is not valid UnknownLabel is raised.
+        If the provided node is not known UnknownNode is raised.
+        """
         if label not in self.BUCKETS:
             raise self.UnknownLabel(f'{label} is not a valid node label')
-
-        if self._nodes.get(node):
-            # Remove any existing labels...
-            for _label in self.BUCKETS:
-                if node in self._marked[_label]:
-                    self._marked[_label].remove(node)
-            # Set the new label
-            self._marked[label].append(node)
+        if self.__nodes.get(node.checksum_address):
+            self.unlabel(node=node)            # Remove any existing labels...
+            self.__marked[label].append(node)  # Set the new label
         else:
-            raise self.UnknownNode(f'Cannot mark an unknown node ({node}).')
-
-    def prune_nodes(self, pruning_strategies: Dict[List[Callable]]) -> None:
-        """Apply node pruning strategies to known nodes once"""
-        for bucket, strategies in pruning_strategies.items():
-            for node in self.get_nodes(label=bucket):
-                for strategy in strategies:
-                    keep = strategy(node=node)
-                    if not keep:
-                        del self[node.checksum_address]    # prune node
-                        self._marked[bucket].remove(node)  # prune corresponding label
-                        # TODO: Trash can label?
-                        break  # this node is already doomed anyways
+            raise self.UnknownNode(f'Cannot label an unknown node ({node}).')
+
+    def unlabel(self, node: "Teacher", label=None) -> None:
+        """Removes a label from a node, or if label is None, all labels are removed."""
+
+        # Remove one label
+        if label:
+            if node in self.__marked[label]:
+                self.__marked[label].remove(node)
+                return
+
+        # Remove all labels
+        for _label in self.BUCKETS:
+            if node in self.__marked[_label]:
+                self.__marked[_label].remove(node)
```

Comment:
A node can't be labelled more than once, so can we short-circuit here and return after removing the node?

Reply:
We can... although it might be nice to have this escape hatch in case a node does somehow end up with two labels?  Am I too paranoid :sweat_smile: ?

Reply:
Additional tests may help reduce to that level of paranoia 😏, but if we keep the "escape hatch" then we should log a warning so that we are at least aware that multi-labelling is occurring, and can subsequently investigate why that is.

Reply:
Logged a warning for now.